### PR TITLE
Adding concatV2 api and Tensor indexing

### PR DESCRIPTION
**Pull request by pschuh**
Using Raw.stridedSlice allows avoiding any of the computation of tensor indexes. Using concat allows users to write code without using the slower slice assignment. This is to improve the perf of the following snippet:

```
func bc_matmul(_ a:Tensor<Float>, _ b:Tensor<Float>) -> Tensor<Float>{
  var res: [Tensor<Float>] = []
  for i in 0..<a.shape[0]{
    res.append((a[i].expandingShape(at: 1) * b).sum(alongAxes: 0))
  }
  return Tensor<Float>(concatenating: res)
}
```

### CODE REVIEW

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow macos

File: stdlib/public/TensorFlow/Ops.swift [hunk 1]
```diff
@@ -686,6 +686,11 @@ public extension Tensor {
   static func ++ (lhs: Tensor, rhs: Tensor) -> Tensor {
     return lhs.concatenated(with: rhs)
   }
+
+  /// Concat N tensors together into a single result tensor.
+  init(concat tensors: [Tensor<Scalar>]) {
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Ops.swift]:
```suggestion
  init(concatenating tensors: [Tensor<Scalar>]) {
```

File: stdlib/public/TensorFlow/Ops.swift [hunk 2]
```diff
@@ -686,6 +686,11 @@ public extension Tensor {
   static func ++ (lhs: Tensor, rhs: Tensor) -> Tensor {
     return lhs.concatenated(with: rhs)
   }
+
+  /// Concat N tensors together into a single result tensor.
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Ops.swift]:
```suggestion
  /// Returns a concatenated tensor of the given tensors.
```

**Review by dan-zheng — APPROVED** [review]
Goodbye to GPE era indexing!

---

Confirmed that [`test/TensorFlowRuntime/tensor.swift`](https://github.com/apple/swift/blob/tensorflow/test/TensorFlowRuntime/tensor.swift) contains element/slice indexing tests.

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
From you PR description:
> a.tensorGet(i)

What is ‘tensorGet(_:)’? For the quality of the PR description, it would be good if you can show an example that actually compiles.

File: stdlib/public/TensorFlow/Ops.swift [hunk 3]
```diff
@@ -686,6 +686,11 @@ public extension Tensor {
   static func ++ (lhs: Tensor, rhs: Tensor) -> Tensor {
     return lhs.concatenated(with: rhs)
   }
+
+  /// Returns a concatenated tensor of the given tensors.
+  init(concatenating tensors: [Tensor<Scalar>]) {
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Ops.swift]: For generality, this should take an array of indices (if concatV2 takes multidimensional indices) with label ‘alongAxes:’, defaulted to 0.

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by pschuh** [comment]
@swift-ci please test tensorflow

**Comment by dan-zheng** [comment]
@pschuh: could you please share some benchmarking results regarding the old/new indexing implementations?

**Pull request merged**